ext/uri/php_uri.c

```diff
@@ -40,12 +48,565 @@ static const zend_module_dep uri_deps[] = {
+	ZEND_ASSERT(uri_object != NULL);
+
+	if (UNEXPECTED(uri_object->internal.uri != NULL)) {
+		uri_object->internal.handler->free_uri(uri_object->internal.uri);
+		uri_object->internal.handler = NULL;
+		uri_object->internal.uri = NULL;
+	}
+
+	zend_object_std_dtor(&uri_object->std);
+}
+
+static zend_object *uri_clone_obj_handler(zend_object *object)
+{
+	uri_object_t *uri_object = uri_object_from_obj(object);
+	uri_internal_t *internal_uri = uri_internal_from_obj(object);
+
+	zend_object *new_object = uri_create_object_handler(object->ce);
+	ZEND_ASSERT(new_object != NULL);
+	uri_object_t *new_uri_object = uri_object_from_obj(new_object);
+
+	URI_ASSERT_INITIALIZATION(internal_uri);
+
+	new_uri_object->internal.handler = internal_uri->handler;
+
+	void *uri = internal_uri->handler->clone_uri(internal_uri->uri);
+	ZEND_ASSERT(uri != NULL);
+
+	new_uri_object->internal.uri = uri;
+
+	zend_objects_clone_members(&new_uri_object->std, &uri_object->std);
+
+	return &new_uri_object->std;
+}
+
+PHPAPI void php_uri_implementation_set_object_handlers(zend_class_entry *ce, zend_object_handlers *object_handlers)
+{
+	ce->create_object = uri_create_object_handler;
+	ce->default_object_handlers = object_handlers;
+	memcpy(object_handlers, &std_object_handlers, sizeof(zend_object_handlers));
+	object_handlers->offset = XtOffsetOf(uri_object_t, std);
+	object_handlers->free_obj = uri_free_obj_handler;
+	object_handlers->clone_obj = uri_clone_obj_handler;
+}
+
+zend_result uri_handler_register(const uri_handler_t *uri_handler)
+{
+	zend_string *key = zend_string_init_interned(uri_handler->name, strlen(uri_handler->name), 1);
+
+	ZEND_ASSERT(uri_handler->name != NULL);
+	ZEND_ASSERT(uri_handler->init_parser != NULL);
+	ZEND_ASSERT(uri_handler->parse_uri != NULL);
+	ZEND_ASSERT(uri_handler->create_invalid_uri_exception != NULL);
+	ZEND_ASSERT(uri_handler->clone_uri != NULL);
+	ZEND_ASSERT(uri_handler->uri_to_string != NULL);
+	ZEND_ASSERT(uri_handler->free_uri != NULL);
+	ZEND_ASSERT(uri_handler->destroy_parser != NULL);
+	ZEND_ASSERT(uri_handler->property_handlers != NULL);
+
+	return zend_hash_add_ptr(&uri_handlers, key, (void *) uri_handler) ? SUCCESS : FAILURE;
+}
 
 static PHP_MINIT_FUNCTION(uri)
 {
+	whatwg_url_ce = register_class_Uri_WhatWg_Url();
+	php_uri_implementation_set_object_handlers(whatwg_url_ce, &whatwg_uri_object_handlers);
+
+	uri_comparison_mode_ce = register_class_Uri_UriComparisonMode();
 	uri_exception_ce = register_class_Uri_UriException(zend_ce_exception);
 	invalid_uri_exception_ce = register_class_Uri_InvalidUriException(uri_exception_ce);
 	whatwg_invalid_url_exception_ce = register_class_Uri_WhatWg_InvalidUrlException(invalid_uri_exception_ce);
+	whatwg_url_validation_error_ce = register_class_Uri_WhatWg_UrlValidationError();
+	whatwg_url_validation_error_type_ce = register_class_Uri_WhatWg_UrlValidationErrorType();
+
+	zend_hash_init(&uri_handlers, 4, NULL, ZVAL_PTR_DTOR, true);
```

Comment:
Is `ZVAL_PTR_DTOR` the correct destructor? Should it perhaps rather be something that calls `handler->destroy_parser()`?

Reply:
With https://github.com/php/php-src/pull/18672#discussion_r2115406859 I agree that the destructor should just be `NULL`.

Resolving this thread.